bpf/bpf_host.c

```diff
@@ -968,6 +976,21 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 __section("from-netdev")
 int from_netdev(struct __ctx_buff *ctx)
 {
+	__u32 __maybe_unused vlan_id;
+
+	/* Filter allowed vlan id's and pass them back to kernel.
+	 */
+	if (ctx->vlan_present) {
+		vlan_id = ctx->vlan_tci & 0xfff;
+		if (vlan_id) {
+			if (allow_vlan(ctx->ifindex, vlan_id))
```

Comment:
If I follow correctly, by default during upgrade, the new `--vlan-bpf-bypass` will be empty, which means that we will generate something like `#define VLAN_FILTER(x, y) { return false; }`. From your recent update on the main PR text, it sounds like this was the cause of your connectivity issue with your apiserver.

This sounds like the new default can break some connectivity in the network, which is not OK. That suggests that this entire new section needs to be conditionally run iff `--vlan-bpf-bypass` is configured.

Reply:
Basically the "drop unconfigured VLANs" logic is OK if and only if there is no way for VLAN traffic to successfully reach its destination today. If some VLAN traffic may reach its destination but other VLAN traffic may end up malformed somewhere else on the network, then changing this code to drop VLAN-tagged traffic by default could cause existing setups to break. We don't want to change this behaviour and arbitrarily break some setups with the default configuration.

Reply:
Sorry for my bad english...

`--vlan-bpf-bypass` will be empty by default, but empty does not mean 'block everythings'. Empty allows vlan id's used by devices controlled by cilium.

Reply:
I worry that we are making this more complex than necessary for the default case to support this specific use case, and I'm already having trouble understanding the implications of the default behaviour.

The lowest-risk option here is to make no changes at all by default. So empty means do not try any special handling for the VLANs. Don't drop, don't bypass. Then separately, if you want "bypass for assigned devices" and/or "drop VLAN-tagged traffic coming in if it's not recognized or part of a list" then you can enable this shiny new option.

Reply:
Just to clarify: current PR logic have no known for me issues.

Reply:
@kvaster Are you configuring the new option? What about if you don't configure the new option?

Reply:
@joestringer, my PR will always allow vlan id's used by devices controlled by cilium - by default traffic will be not blocked and will be controlled by cilium. You need to configure this option only if you have additional vlan devices which are not controlled by cilium (have bpf programs attached).

Reply:
That means defaults with this PR (without configuring new option) should be enough for all reported cases in issue and should not break security.

Reply:
> my PR will always allow vlan id's used by devices controlled by cilium - by default traffic will be not blocked and will be controlled by cilium. 

When you say "allow" here, do you mean `return CTX_ACT_OK`, ie "bypass Cilium logic"? That's the security concern I had in the previous issue that I was looking to avoid changing by default.

Reply:
Yes, I mean `return CTX_ACT_OK`. And it will not bypass cilium logic, cause in this case packet will be processed by cilium by bpf program attached to vlan device - only vlan id's that will be processed by cilium are allowed by default... Generally we just don't care if we have additional devices and bpf programs are not attached to them. To make it proper we will have to rewrite vlan logic in bpf, but personally I think this is too much. Without such changes cilium will malform vlan tagged packets when attached to main device (to which vlan devices attached), cause it should not process such packets. And blocking every vlan tagged packet will lead to broken connectivity without possibility to investigate problem - probably this will happen only on first setup before users will know that they should explicitly allow vlan bpf bypass, but it will be hard to understand what is happening. I think that somebody from cilium team should decide what kind of behaviour is OK. I don't understand what can be done more to fix this issue properly. Thanks.

Reply:
> And it will not bypass cilium logic, cause in this case packet will be processed by cilium by bpf program attached to vlan device 

I think you're basically saying, this logic is only bypassing Cilium BPF if it can identify that there is a VLAN device behind which will handle the traffic. I missed the nuance there initially, and at least in principle I agree that this mitigates the security concern we raised earlier since we have basically made a check to ensure that the traffic will be handled by Cilium after we skip processing.

My next question is, how do we know that the Cilium BPF program attached to the VLAN device is actually handling the traffic? Today the change is implemented through directly modifying `from_netdev()` / `to_netdev()` which is presumably used for both the BPF program on the main device and also on the VLAN device. Does this work because `ctx->vlan_present` is true only for VLAN packets arriving on the main device and it's false for decapsulated packets arriving on the vlan device?

Reply:
Yes, it will work cause kernel will strip vlan tag and will process packet on vlan device - and bpf will process such packet cause vlan_present will be false already.

Reply:
Also please note that allow_vlan function have ifindex argument. And if, for example, we will receive packet with double vlan tags then first one will be stripped, packet will be redirected for processing on vlan device and will be dropped there by allow_vlan function, cause no vlan tag is allowed on vlan device.